atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientPool.java

```diff
@@ -0,0 +1,428 @@
+            for (List<InetSocketAddress> rangeOwners : tokenMap.asMapOfRanges().values()) {
+                for (InetSocketAddress address : rangeOwners) {
+                    serversToAdd.add(address);
+                }
+            }
+        }
+        Sets.difference(Sets.difference(serversToAdd, currentPools.keySet()), blacklistedHosts.keySet());
+
+        if (!config.autoRefreshNodes()) { // (we would just add them back in)
+            serversToRemove = Sets.difference(currentPools.keySet(), config.servers());
+        }
+
+        for (InetSocketAddress newServer : serversToAdd) {
+            currentPools.put(newServer, new CassandraClientPoolingContainer(newServer, config));
+        }
+
+        if (!(serversToAdd.isEmpty() && serversToRemove.isEmpty())) { // if we made any changes
+            sanityCheckRingConsistency();
+            if(!config.autoRefreshNodes()) { // grab new token mapping, if we didn't already do this before
+                refreshTokenRanges();
+            }
+        }
+    }
+
+    private void checkAndUpdateBlacklist() {
+        // Check blacklist and re-integrate or continue to wait as necessary
+        for (Map.Entry<InetSocketAddress, Long> blacklistedEntry : blacklistedHosts.entrySet()) {
+            if (blacklistedEntry.getValue() + BANNED_HOST_BACKOFF_TIME_MS < System.currentTimeMillis()) {
+                InetSocketAddress host = blacklistedEntry.getKey();
+                if(isHostHealthy(host)) {
+                    blacklistedHosts.remove(host);
+                    log.error("Added host %s back into the pool after a waiting period and successful health check.", host);
+                }
+            }
+        }
+    }
+
+    private void addToBlacklist(InetSocketAddress badHost) {
+        blacklistedHosts.put(badHost, System.currentTimeMillis());
+    }
+
+    private boolean isHostHealthy(InetSocketAddress host) {
+        try {
+            CassandraClientPoolingContainer testingContainer = new CassandraClientPoolingContainer(host, config);
+            testingContainer.runWithPooledResource(describeRing);
+        } catch (Exception e) {
+            log.error("We tried to add %s back into the pool, but got an exception that caused to us distrust this host further.", host, e);
+            return false;
+        }
+        return true;
+    }
+
+    private CassandraClientPoolingContainer getRandomGoodHost() {
+        Set<InetSocketAddress> livingHosts = Sets.difference(currentPools.keySet(), blacklistedHosts.keySet());
+        if (livingHosts.isEmpty()) {
+            log.error("There are no known live hosts in the connection pool. We're choosing one at random in a last-ditch attempt at forward progress.");
+            livingHosts = currentPools.keySet();
+        }
+
+        return currentPools.get(livingHosts.toArray(new InetSocketAddress[0])[ThreadLocalRandom.current().nextInt(currentPools.size())]);
```

Comment:
I think we'll need something along the lines of the following to handle the case where `currentPools` is being refreshed concurrently:

```
    private CassandraClientPoolingContainer getRandomGoodHost() {
        Map<InetSocketAddress, CassandraClientPoolingContainer> pools = currentPools; // volatile read
        Set<InetSocketAddress> livingHosts = Sets.difference(pools.keySet(), blacklistedHosts.keySet());
        if (livingHosts.isEmpty()) {
            log.error("There are no known live hosts in the connection pool. We're choosing one at random in a last-ditch attempt at forward progress.");
            livingHosts = pools.keySet();
        }

        InetSocketAddress[] liveHostAddresses = livingHosts.toArray(new InetSocketAddress[0]);
        return pools.get(liveHostAddresses[ThreadLocalRandom.current().nextInt(liveHostAddresses.length)]);
    }
```